src/Microsoft.AspNet.Mvc.TagHelpers/InputTagHelper.cs

```diff
@@ -236,25 +236,27 @@ private void GenerateCheckBox(ModelExplorer modelExplorer, TagHelperOutput outpu
                 }
             }
 
-            var tagBuilder = Generator.GenerateCheckBox(
+            var checkBoxTag = Generator.GenerateCheckBox(
                 ViewContext,
                 modelExplorer,
                 For.Name,
                 isChecked: null,
                 htmlAttributes: htmlAttributes);
-            if (tagBuilder != null)
+            if (checkBoxTag != null)
             {
                 // Do not generate current element's attributes or tags. Instead put both <input type="checkbox"/> and
                 // <input type="hidden"/> into the output's Content.
                 output.Attributes.Clear();
                 output.TagName = null;
 
-                output.Content.Append(tagBuilder.ToHtmlContent(TagRenderMode.SelfClosing));
+                checkBoxTag.TagRenderMode = TagRenderMode.SelfClosing;
```

Comment:
Instead set `TagRenderMode` when creating the `TagBuilder` instance. Current separation between `TagBuilder` content and whether or not the element should be self-closing is artificial, just a result of the previous APIs. Might even be worth adding the mode or a `bool isSelfClosing` parameter to a `TagBuilder` constructor overload, saving the property setter for the (rare) `StartTag` / `EndTag` special cases.

Here for example `DefaultHtmlGenerator` knows the `<input>` tag is self-closing and this code shouldn't care.